### ISSUE TITLE
Nested Lists - Getting the first value

### ISSUE THREAD

**Issue by pdaniele** [STAT545-UBC/Discussion#470]
Hello!

I'm working with the ice and fire API gathering data on characters, houses and books. I'm looking to create a dataset that is one row per character with many of their attributes. I'm running into a problem when working with nested lists. 

I have a list called characters, which contains their attributes, and mapping elements with only one value works well using the code below.
map_df(characters, `[`, c('name', 'gender', 'culture', 'born', 'died', 'father', 'mother', 'spouse'))

However, when I try to map (in all sorts of ways) to an element of the list which is also a list such as allegiances I can't seem to map those values into a data frame with purrr functions. To simplify things I just want whatever the first value of the sub-list is. However, I need the result to be on the same line as the character's name or at least have an identifier which will allow me to join to it. The following for loop works, but seems rather clunky.  

```{r}
allegiances <- data.frame()
for(i in 1:10) {
  
## Create character ID Variable
  allegiances[i,1] <- i

## If length is zero, there are no entries. Set it to NULL
  if(length(characters2[[i]][[12]]) == 0) {
    allegiances[i,2] <- 'NULL'
  }
## If the length is not zero, there are entries! Take the first!
  if(length(characters2[[i]][[12]]) != 0) {
    x<- characters2[[i]][[12]][[1]]
    allegiances[i,2] <- substr(x, start=45, stop=(45 +nchar(x)-45)) 
  }
}
```

Any tips or resources I could check out? Hopefully this all makes sense.

**Comment by samhinshaw** [comment]
Absolutely! Check out `magrittr::extract()` and that suite of functions. 

>magrittr provides a series of aliases which can be more pleasant to use when composing chains using the %>% operator

I think that will help your code flow much smoother, which may in turn help you sort out the specific issues you're having! If not, feel free to post a follow-up and I'll take a look. 

If you can provide some sample data or the API request you used to form the data.frame that'd help.

**Comment by pdaniele** [comment]
Here is the API request I used, in case it helps.

```{r}
example <- GET('http://www.anapioficeandfire.com/api/characters/?page=1&pageSize=50')
example_parsed <- content(example, as='parsed')
```

Which gives me a list of 50 characters, containing a mix of character elements and lists. Using the code below, I can get the character elements into a dataframe, but I also need to get at least 1 of the allegiances (which are in a list, and the list isn't guaranteed to have any elements) into a dataframe. 

```{r}
characters <- map_df(example_parsed , `[`, c('url','name', 'gender', 'culture', 'born', 'died'))
```
My end goal is one row per person. So I need a data structure that is something like:
URL | Allegiances 
Which I can then join onto my other dataset later
or
 URL | Name | Gender | Culture | Born | Died | Allegiances (First in list)
which saves me the joining.

Thanks for your help. I'll take a look at the function you recommended in the mean time and see if I can figure it out.

**Comment by jennybc** [comment]
This should get you started:

``` r
library(httr)
library(tidyverse)
#> Loading tidyverse: ggplot2
#> Loading tidyverse: tibble
#> Loading tidyverse: tidyr
#> Loading tidyverse: readr
#> Loading tidyverse: purrr
#> Loading tidyverse: dplyr
#> Conflicts with tidy packages ----------------------------------------------
#> filter(): dplyr, stats
#> lag():    dplyr, stats

example <- GET('http://www.anapioficeandfire.com/api/characters/?page=1&pageSize=50')
example_parsed <- content(example, as='parsed')
characters <- map_df(example_parsed , `[`, 
                     c('url','name', 'gender', 'culture', 'born', 'died'))
allegiances <- example_parsed %>% 
  tibble(
    name = map_chr(., "name"),
    allegiances = map(., "allegiances")
  )
x <- left_join(characters, allegiances)
#> Joining, by = "name"
```

This is complicated by your desire to keep `allegiances`. I recently opened an issue about `map_df()` being unable to create list-columns. Maybe that can happen? But no soon enough for you!

You might also want to look at my data prep for the repurrrsive package, which does the next step of resolving which houses are being referred to in `allegiances`.

https://github.com/jennybc/repurrrsive/blob/master/data-raw/got_pov-chars-wrangle.R

**Comment by pdaniele** [comment]
Thank you both your help. I was able to swap out the for loops with your code. So much nicer!
